### PR TITLE
Include informative names in-this-file element

### PR DESCRIPTION
**Pull request by FadhlanR**
Ticket: CS-5993

Screenshot 2023-10-17 at 16 29 41

### CODE REVIEW

**Comment by habdelra** [comment]
i'm confused, i thought a big part of this is to be able to describe non-exported cards in human language, like: "Address Field ( fieldOf Employee Card, ancestorOf HomeAddress)". I'm not seeing any tests or logic for this kind of thing.

**Comment by tintinthong** [comment]
Not sure this was intended to be included in another PR. Can you follow thru the design fully in this PR

Screenshot 2023-10-18 at 1 39 41 PM

image

File: packages/host/app/components/operator-mode/detail-panel-selector.gts [hunk 1]
```diff
@@ -191,10 +219,27 @@ export default class Selector extends Component<Signature> {
         }
 
         .selector-item {
+          --icon-color: var(--boxel-highlight);
+
           display: flex;
           align-items: center;
+          flex-wrap: wrap;
+          text-wrap: nowrap;
```
Comment (tintinthong) [inline packages/host/app/components/operator-mode/detail-panel-selector.gts]: can you follow the css as in recent files so user knows when text exist. 

```
        overflow-wrap: anywhere;
        overflow: hidden;
```

I think this has been the overflow pattern for LHS. There was a recommendation to use ellipsis as the text-overflow but I think we need to set some max-width for the container. I think for now using the css above shud be fine so at least its consistent in the top (detail-panel) and bottom (recent-files)

**Comment by tintinthong** [comment]
I notice as part of comment above that there is no distinction between `function` or `class` currently. You'd probably need to add some `type` property in the declarations in the schema analysis plugin to distinguish a function from class. In addition, you may be able to add a `type` property for cards or fields also, but I wonder if its actually that straightforward (Note there is a reason why its possibleFields and possibleCardsOrFields). How I check for card or field is shown in the `isCard` or `isField` function inside the `detail-panel`.

**Comment by FadhlanR** [comment]
> I notice as part of comment above that there is no distinction between `function` or `class` currently. You'd probably need to add some `type` property in the declarations in the schema analysis plugin to distinguish a function from class. In addition, you may be able to add a `type` property for cards or fields also, but I wonder if its actually that straightforward (Note there is a reason why its possibleFields and possibleCardsOrFields). How I check for card or field is shown in the `isCard` or `isField` function inside the `detail-panel`.

@tintinthong, I decided to add the `type` property in `BaseDeclarations` to determine whether it's a function or class because in the `schema-analysis-plugin`, we lack sufficient data to distinguish `card` from `field`. So, I also added another function in `detail-panel-selector` to identify whether it's a `card`, `field`, `class`, or `function.
Screenshot 2023-10-18 at 17 55 02

**Comment by FadhlanR** [comment]
> i'm confused, i thought a big part of this is to be able to describe non-exported cards in human language, like: "Address Field ( fieldOf Employee Card, ancestorOf HomeAddress)". I'm not seeing any tests or logic for this kind of thing.

It seems we decided not to have that in our design. https://linear.app/cardstack/issue/CS-5993/include-informative-names-in-this-file-element#comment-dde003c3

File: packages/host/app/components/operator-mode/detail-panel-selector.gts [hunk 2]
```diff
@@ -105,9 +123,35 @@ export default class Selector extends Component<Signature> {
...
+                          selectorItem.declaration.exportedAs
+                          selectorItem.declaration.localName
+                        )
+                      }}<span
+                        >({{selectorItem.declaration.localName}})</span>{{/unless}}
+                    {{else}}
+                      <span class='non-exported'>{{if
+                          selectorItem.declaration.localName
+                          selectorItem.declaration.localName
+                          '??'
```
Comment (burieberry) [inline packages/host/app/components/operator-mode/detail-panel-selector.gts]: Was the `??` a placeholder? Is there a case where there's no name?

**Review by burieberry — COMMENTED** [review]
Looks good. Made a comment in the code. Nice icon import without svg-jar.

File: packages/host/app/components/operator-mode/detail-panel-selector.gts [hunk 3]
```diff
@@ -73,6 +83,14 @@ export default class Selector extends Component<Signature> {
...
   }
 
+  getType(declaration: ModuleDeclaration) {
+    let type = declaration.type as string;
+    if (isCardOrFieldDeclaration(declaration)) {
+      type = isCardDef(declaration.cardOrField) ? 'card' : 'field';
+    }
+    return type;
+  }
+
```
Comment (tintinthong) [inline packages/host/app/components/operator-mode/detail-panel-selector.gts:93]: 👍

File: packages/host/app/components/operator-mode/detail-panel-selector.gts (hunk 2 above)
Comment (tintinthong) [reply packages/host/app/components/operator-mode/detail-panel-selector.gts]: Yea we should replace this with something that makes sense. [No Name Found] as previously

**Review by tintinthong — APPROVED** [review]
LGTM. Address the issue of `??` as burcu mentioned

**Pull request merged**